### PR TITLE
Tweak: Deprecations - Need to update the text of the deprecation notification [ED-9801]

### PR DESCRIPTION
**Pull request by confucius-elementor**
https://elementor.atlassian.net/browse/ED-9801

### CODE REVIEW

File: modules/dev-tools/deprecation.php [hunk 1]
```diff
@@ -123,7 +123,7 @@ public function parse_version( $version ) {
 		$version_explode_count = count( $version_explode );
 
 		if ( $version_explode_count < 3 || $version_explode_count > 4 ) {
-			trigger_error( 'Invalid Semantic Version string provided' );
+			trigger_error( 'Invalid Semantic Version string provided' . $version );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:126]: It's better to `dump` it in case it's null or empty string.
```suggestion
			trigger_error( 'Invalid Semantic Version string provided' . var_export( $version, 1 ) );
```

File: modules/dev-tools/deprecation.php [hunk 2]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+		if ( $this->should_print_deprecated( $version, $base_version ) || $this->should_console_log_deprecated() ) {
+			$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
+
+			$external_sources = $this->find_third_party_sources( $backtrace );
+
+			$source_message = 'Elementor';
+			$plugin_name = 'unknown';
+			if ( ! ( empty( $external_sources ) ) ) {
+				$calling_source = $external_sources[ array_key_first( $external_sources ) ];
+				$plugin_name = $this->get_plugin_name( $calling_source['file'] );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:213]: Maybe move to `find_third_party_sources`

File: modules/dev-tools/deprecation.php [hunk 3]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+				if ( empty( $plugin_name ) ) {
+					$plugin_name = 'none';
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:215]: `none` or `unknown`?

File: modules/dev-tools/deprecation.php [hunk 4]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+				}
 			}
 
-			if ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG ) {
-				$print_deprecated = true;
+			if ( $this->should_print_deprecated( $version, $base_version ) ) {
+				$print_deprecated = array(
+					'source' => $source_message,
+					'plugin' => $plugin_name,
+				);
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:234]: Use short array `[]`

File: modules/dev-tools/deprecation.php [hunk 5]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+			$plugin_name = 'unknown';
+			if ( ! ( empty( $external_sources ) ) ) {
+				$calling_source = $external_sources[ array_key_first( $external_sources ) ];
+				$plugin_name = $this->get_plugin_name( $calling_source['file'] );
+				if ( empty( $plugin_name ) ) {
+					$plugin_name = 'none';
+				}
+				$source_message = sprintf( '%s on file %s:%d.', $calling_source['function'], $calling_source['file'], $calling_source['line'] );
+			}
+			if ( $this->should_console_log_deprecated() ) {
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:219]:
```suggestion
			}

			if ( $this->should_console_log_deprecated() ) {
```

File: modules/dev-tools/deprecation.php [hunk 6]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+					'plugin' => $plugin_name,
+				);
 			}
 		}
-
 		return $print_deprecated;
 	}
 
+	private function get_plugin_name( $filename ) {
+		$plugin_path = substr( strstr( $filename, WP_PLUGIN_DIR ), strlen( WP_PLUGIN_DIR ) );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:241]: It can be a `theme`, or `mu-plugin` so try to work with `WP_CONTENT_DIR`

File: modules/dev-tools/deprecation.php [hunk 7]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+		if ( empty( $plugin_path ) ) {
+			return '';
+		}
+		$plugin_name = explode( '/', $plugin_path )[1];
+		return $plugin_name;
+	}
+
+	private function is_elementor_file( $stack_element ) {
+		$filename = $stack_element['file'];
+		return ( strpos( $filename, 'elementor/' ) !== false || strpos( $filename, 'elementor-pro/' ) !== false );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:251]: It can be 'something-for-elementor`, please check more strict

File: modules/dev-tools/deprecation.php [hunk 8]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+	 * If the user is logged in but does not have admin privileges, the message will be printed only if the user has ELEMENTOR_DEBUG enabled.
+	 *
+	 *
+	 * @param  mixed $version
+	 * @param  mixed $base_version
+	 * @return bool
+	 */
+	private function should_print_deprecated( $version, $base_version ) {
+
+		$elementor_debug = ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:283]:
```suggestion
		$is_elementor_debug = ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG );
```

File: modules/dev-tools/deprecation.php [hunk 9]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+		$wp_debug = ( defined( 'WP_DEBUG' ) && WP_DEBUG );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:284]:
```suggestion
		$is_wp_debug = ( defined( 'WP_DEBUG' ) && WP_DEBUG );
```

File: modules/dev-tools/deprecation.php [hunk 10]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+
+		return false;
+
+	}
+
+	private function should_console_log_deprecated() {
+		return ( defined( 'WP_DEBUG' ) && WP_DEBUG );
+	}
+
+	private function notify_deprecated_function( $function_name, $version, $replacement = '', $calling_plugin = '', $call_location = '' ) {
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:313]:
```suggestion
	// Extends WordPress `_deprecated_function` with plugin info.
	private function notify_deprecated_function( $function_name, $version, $replacement = '', $calling_plugin = '', $call_location = '' ) {
```

File: modules/dev-tools/deprecation.php [hunk 11]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:314]: PHP Lint: A method/function should not have a first empty line

File: modules/dev-tools/deprecation.php [hunk 12]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+					$message_string, // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+					$error_message_args // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+				),
+				E_USER_DEPRECATED
+			);
+		}
+	}
+
+
+
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:356]: PHP Lint: Remove unnecessary empty lines

File: modules/dev-tools/deprecation.php [hunk 13]
```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+	private function notify_deprecated_argument( $argument, $version, $replacement = '', $message = '', $plugin, $source ) {
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:357]: Please rewrite based on WordPress `_deprecated_argument`

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php [hunk 14]
```diff
@@ -196,9 +196,59 @@ public function test_deprecated_function_soft() {
...
 			'0.0.0',
 			'',
+			'Elementor',
 		], $settings['soft_notices']['test_deprecated_function_soft'] );
 	}
 
+
+	public function test_deprecated_function_as_admin() {
+		$this->act_as_admin();
+		$function_name = 'blablabla';
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:206]:
```suggestion
		$function_name = 'test';
```

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php [hunk 15]
```diff
@@ -196,9 +196,59 @@ public function test_deprecated_function_soft() {
...
+		$this->setExpectedDeprecated( $function_name );
+		$this->run_function_deprecated_test( $function_name );
+	}
+
+	public function test_deprecated_argument_as_admin() {
+		$this->act_as_admin();
+		$argument_name = 'argumatura';
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:213]:
```suggestion
		$argument_name = 'test';
```

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php [hunk 16]
```diff
@@ -196,9 +196,59 @@ public function test_deprecated_function_soft() {
...
+		$this->expectError();
+		$this->run_argument_deprecated_test( $argument_name );
+	}
+
+	public function test_deprecated_function_not_as_admin_does_not_show_message() {
+		$this->run_function_deprecated_test();
+		$this->assertTrue(true); // hack because phpunit demands assert
+	}
+
+	private function run_function_deprecated_test($function_name = 'blablabla'){
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:224]: Always prefer to use `test` as test data
```suggestion
	private function run_function_deprecated_test($function_name = 'blablabla'){
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/dev-tools/deprecation.php (hunk 2 above)
Comment (confucius-elementor) [reply modules/dev-tools/deprecation.php:213]: I'd rather not, prefer to keep find_third_party_sources() as pure and designated as possible

File: modules/dev-tools/deprecation.php (hunk 6 above)
Comment (confucius-elementor) [reply modules/dev-tools/deprecation.php:241]: i tried, it didn't catch the plufins

File: modules/dev-tools/deprecation.php (hunk 8 above)
Comment (confucius-elementor) [reply modules/dev-tools/deprecation.php:283]: sorry, I disagree. it's exactly elementor_debug, the value of the "define", just undefined-safe

File: modules/dev-tools/deprecation.php (hunk 9 above)
Comment (confucius-elementor) [reply modules/dev-tools/deprecation.php:284]: same here

File: modules/dev-tools/deprecation.php [hunk 17]
```diff
@@ -123,7 +123,7 @@ public function parse_version( $version ) {
 		$version_explode_count = count( $version_explode );
 
 		if ( $version_explode_count < 3 || $version_explode_count > 4 ) {
-			trigger_error( 'Invalid Semantic Version string provided' );
+			trigger_error( 'Invalid Semantic Version string provided' . var_export( $version, 1 ) );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:126]:
```suggestion
			trigger_error( 'Invalid Semantic Version string provided: ' . var_export( $version, 1 ) );
```

File: modules/dev-tools/deprecation.php [hunk 18]
```diff
@@ -199,29 +199,194 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
 			throw new \Exception( 'Invalid deprecation diff.' );
 		}
 
-		$print_deprecated = false;
+		$print_deprecated = array();
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:202]:
```suggestion
		$print_deprecated = [];
```

File: modules/dev-tools/deprecation.php [hunk 19]
```diff
@@ -199,29 +199,194 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+		if ( strpos( $filename, WP_PLUGIN_DIR ) !== false ) {
+			return true;
+		}
+		return false;
+
+	}
 
-			if ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG ) {
-				$print_deprecated = true;
+	private function find_third_party_sources( $stack_trace ) {
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:264]:
```suggestion
	private function filter_third_party_sources( $stack_trace ) {
```

File: modules/dev-tools/deprecation.php [hunk 20]
```diff
@@ -199,29 +199,194 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
-				$this->soft_deprecated_notices[ $entity ] = [
-					$version,
-					$replacement,
+		if ( $this->should_print_deprecated( $version, $base_version ) || $this->should_console_log_deprecated() ) {
+			$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
+
+			$external_sources = $this->find_third_party_sources( $backtrace );
+
+			$source_message = 'Elementor';
+			$plugin_name = 'unknown';
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:210]:
```suggestion
			$plugin_name = 'Unknown';
```

File: modules/dev-tools/deprecation.php (hunk 6 above)
Comment (matipojo) [reply modules/dev-tools/deprecation.php:242]: https://github.com/elementor/elementor/pull/22062/files#diff-d2f15ad4308ce3f20df282bb3817b49c5f24aef0dddf8f818a791e8bea3d92bdR401

File: modules/dev-tools/deprecation.php [hunk 21]
```diff
@@ -199,29 +199,194 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
+		if ( empty( $plugin_path ) ) {
+			return '';
+		}
+		$plugin_name = explode( '/', $plugin_path )[1];
+		return $plugin_name;
+	}
+
+	private function is_elementor_file( $stack_element ) {
+		$filename = $stack_element['file'];
+		return ( strpos( $filename, '//elementor/' ) !== false || strpos( $filename, 'elementor-pro/' ) !== false );
```
Comment (matipojo) [inline modules/dev-tools/deprecation.php:252]:
```suggestion
		return ( strpos( $filename, '//elementor/' ) !== false || strpos( $filename, '//elementor-pro/' ) !== false );
```

**Review by matipojo — DISMISSED** [review]

File: composer.json [hunk 22]
```diff
@@ -19,8 +19,11 @@
 	},
 	"scripts": {
 	  "lint": "phpcs --standard=ruleset.xml",
+	  "lint:help": "phpcs --version",
 	  "test": "phpunit",
 	  "test:install": "bash ./bin/install-wp-tests-local.sh",
+	  "test:m": "phpunit  tests/phpunit/elementor/modules/dev-tools/test-deprecation.php",
+	  "test:m2": "phpunit 'Elementor\\Tests\\Phpunit\\Elementor\\Core\\Common\\DevTools'",
```
Comment (KingYes) [inline composer.json:26]: What is M and M2?

File: modules/dev-tools/deprecation.php [hunk 23]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+	 * If the user is not logged in, the message will be printed only if the user has WP_DEBUG enabled.
+	 * If the user is logged in but does not have admin privileges, the message will be printed only if the user has ELEMENTOR_DEBUG enabled.
+	 *
+	 *
+	 * @param  mixed $version
+	 * @param  mixed $base_version
+	 * @return bool
+	 */
+	private function should_print_deprecated( $version, $base_version ) {
+
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:308]:
```suggestion
```

File: modules/dev-tools/deprecation.php [hunk 24]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+		$filename = $stack_element['file'];
+		if ( strpos( $filename, WP_PLUGIN_DIR ) !== false ) {
+			return true;
+		}
+		return false;
+	}
+
+	private function filter_third_party_sources( $stack_trace ) {
+		$sources = array_filter($stack_trace, function ( $elem ) {
+			return ( ! ( $this->is_elementor_file( ( $elem ) ) ) );
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:291]:
```suggestion
			return ! $this->is_elementor_file( $elem );
```

File: modules/dev-tools/deprecation.php [hunk 25]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+	private function is_plugin( $stack_element ) {
+		$filename = $stack_element['file'];
+		if ( strpos( $filename, WP_PLUGIN_DIR ) !== false ) {
+			return true;
+		}
+		return false;
+	}
+
+	private function filter_third_party_sources( $stack_trace ) {
+		$sources = array_filter($stack_trace, function ( $elem ) {
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:290]:
```suggestion
		$sources = array_filter( $stack_trace, function( $elem ) {
```

File: modules/dev-tools/deprecation.php [hunk 26]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+
+		$print_deprecated = [];
+
+		$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
+		$external_sources = $this->filter_third_party_sources( $backtrace );
+
+		$source_message = 'Elementor';
+		$plugin_name = 'Unknown';
+		$source_type = '';
+		if ( ! ( empty( $external_sources ) ) ) {
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:212]:
```suggestion
		if ( ! empty( $external_sources ) ) {
```

File: modules/dev-tools/deprecation.php [hunk 27]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
-	 * @return bool|void
+	 * @return array
 	 * @throws \Exception
 	 */
-	private function check_deprecation( $entity, $version, $replacement, $base_version = null ) {
+	private function check_deprecation( $version, $base_version = null ) {
 		if ( null === $base_version ) {
 			$base_version = $this->current_version;
 		}
-
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:195]: Why did you remove it?

File: modules/dev-tools/deprecation.php [hunk 28]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+	 * If the user is logged in but does not have admin privileges, the message will be printed only if the user has ELEMENTOR_DEBUG enabled.
+	 *
+	 *
+	 * @param  mixed $version
+	 * @param  mixed $base_version
+	 * @return bool
+	 */
+	private function should_print_deprecated( $version, $base_version ) {
+
+		$is_elementor_debug = ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG );
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:309]:
```suggestion
		$is_elementor_debug = defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG;
```

File: modules/dev-tools/deprecation.php [hunk 29]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+		$is_wp_debug = ( defined( 'WP_DEBUG' ) && WP_DEBUG );
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:310]:
```suggestion
		$is_wp_debug = defined( 'WP_DEBUG' ) && WP_DEBUG;
```

File: modules/dev-tools/deprecation.php [hunk 30]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+				return true;
+			}
+			if ( $user_is_admin ) {
+				return true;
+			}
+			if ( $user_is_logged_in && ! $first_deprecation_stage ) {
+				return true;
+			}
+		}
+		return false;
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:330]:
```suggestion
		if ( ! $is_wp_debug ) {
			return false;
		}

		if ( $is_elementor_debug ) {
			return true;
		}
		
		if ( $user_is_admin ) {
			return true;
		}
		
		if ( $user_is_logged_in && ! $first_deprecation_stage ) {
			return true;
		}

		return false;
```

File: modules/dev-tools/deprecation.php [hunk 31]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+	}
+
+	private function should_console_log_deprecated() {
+		return ( defined( 'WP_DEBUG' ) && WP_DEBUG );
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:334]:
```suggestion
		return defined( 'WP_DEBUG' ) && WP_DEBUG;
```

File: modules/dev-tools/deprecation.php [hunk 32]
```diff
@@ -185,43 +183,234 @@ public function compare_version( $version1, $version2 ) {
...
+	}
 
-			if ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG ) {
-				$print_deprecated = true;
+	private function notify_deprecated_function( $function_name, $version, $replacement = '', $calling_plugin = '', $call_location = '', $type = 'plugin' ) {
```
Comment (KingYes) [inline modules/dev-tools/deprecation.php:337]: A lot of parameters for one method?

**Review by KingYes — DISMISSED** [review]

File: modules/dev-tools/deprecation.php (hunk 27 above)
Comment (confucius-elementor) [reply modules/dev-tools/deprecation.php:195]: Only changed the return type to array.

File: modules/dev-tools/backtrace-helper.php [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+	 * find_who_called_me
+	 * Retrieves the function, class, file, line, type and name of the function that called the function that called this function.
+	 *
+	 * @param  int $stack_depth The depth of the stack to look for.
+	 * @return array
+	 */
+	public static function find_who_called_me( $stack_depth ) {
+		$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
+
+		$caller = $backtrace[ $stack_depth ];
```
Comment (Nevoss) [inline modules/dev-tools/backtrace-helper.php:19]: What if $bactrace is not an array or the `$stack_depth` key is not exists in the array?

File: modules/dev-tools/backtrace-helper.php [hunk 34]
```diff
@@ -0,0 +1,51 @@
...
+			'file' => $caller_file,
+			'line' => $caller_line,
+			'type' => $source['type'],
+			'name' => $source['name'],
+		);
+		return $res;
+	}
+
+	private static function get_source( $filename ) {
+		$file = str_replace( WP_CONTENT_DIR, '', $filename, $is_in_content );
```
Comment (Nevoss) [inline modules/dev-tools/backtrace-helper.php:38]: Try avoid using the 4th param the count

File: modules/dev-tools/backtrace-helper.php [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+		$name = 'Unknown';
+		$type = '';
+		if ( 1 === $is_in_content ) {
+			$root_folder = explode( '/', $file );
+			$type = $root_folder[1];
+			$name = $root_folder[2];
+		}
```
Comment (Nevoss) [inline modules/dev-tools/backtrace-helper.php:45]: What `type` and `name` mean?

```suggestion
			[,$type, $name] = explode( '/', $file );
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: modules/dev-tools/backtrace-helper.php [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+			$caller = $backtrace[ $stack_depth ];
+		}
+
+		$caller_function = $caller['function'] ?? '';
+		$caller_class = $caller['class'] ?? '';
+		$caller_file = $caller['file'] ?? '';
+		$caller_line = $caller['line'] ?? '';
+		$source = self::get_source( $caller_file );
+
+		$res = array(
```
Comment (confucius-elementor) [inline modules/dev-tools/backtrace-helper.php:36]: Use newer syntas: $res = [ ....... ]

File: modules/dev-tools/backtrace-helper.php [hunk 37]
```diff
@@ -0,0 +1,65 @@
...
+	 * line - location
+	 * type - plugin or theme
+	 * name - plugin/theme name
+	 */
+	public static function find_who_called_me( $stack_depth ) {
+		$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
+
+		$caller = [];
+		if ( array_key_exists( $stack_depth, $backtrace ) ) {
+			$caller = $backtrace[ $stack_depth ];
```
Comment (confucius-elementor) [inline modules/dev-tools/backtrace-helper.php:27]: In this case, return null instead of an empty array at the end

File: modules/dev-tools/deprecation.php [hunk 38]
```diff
@@ -178,50 +180,191 @@ public function compare_version( $version1, $version2 ) {
...
+		$deprecation_info = [
+			'entity' => $entity,
+			'version' => $version,
+			'replacement' => $replacement,
+			'calling_function' => $calling_source['function'] ?? '',
+			'calling_file' => $calling_source['file'] ?? '',
+			'calling_line' => $calling_source['line'] ?? '',
+			'calling_plugin' => $calling_source['name'] ?? '',
+			'calling_type' => $calling_source['type'] ?? '',
+		];
```
Comment (Nevoss) [inline modules/dev-tools/deprecation.php:205]:
```suggestion
		$deprecation_info = [
			'entity' => $entity,
			'version' => $version,
			'replacement' => $replacement,
			'calling_source' => $calling_source,
		];
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

**Review by Nevoss — APPROVED** [review]
